src/chronify/time_utils.py

```diff
@@ -0,0 +1,86 @@
+"""Functions related to time"""
+
+import logging
+import numpy as np
+from typing import Any
+
+import pandas as pd
+
+from chronify.time import (
+    TimeIntervalType,
+)
+from chronify.exceptions import InvalidParameter
+
+logger = logging.getLogger(__name__)
+
+
+def shift_time_interval(
+    df: "pd.Series[pd.Timestamp]",
+    from_interval_type: TimeIntervalType,
+    to_interval_type: TimeIntervalType,
+) -> "pd.Series[pd.Timestamp]":
+    """Shift pandas timeseries by time interval based on interval type."""
+    assert (
+        from_interval_type != to_interval_type
+    ), f"from_ and to_interval_type are the same: {from_interval_type}"
+    arr = np.sort(df)
+    freqs = set((np.roll(arr, -1) - arr)[:-1])
+    assert len(freqs), f"Timeseries has more than one frequency, {freqs}"
+    freq: pd.Timedelta = list(freqs)[0]
```

Comment:
Isn't this a numpy type?

Secondly, you can avoid the intermediate list with `next(iter(freqs))`.

Reply:
This is pd.Timedelta.

Reply:
`freq: pd.Timedelta = next(iter(freqs))`?

Reply:
```
-> match (from_interval_type, to_interval_type):
(Pdb) l
 27  	   freqs = set((np.roll(arr, -1) - arr)[:-1])
 28  	   assert len(freqs), f"Timeseries has more than one frequency, {freqs}"
 29  	   freq: pd.Timedelta = next(iter(freqs))
 30  	   breakpoint()
 31  	
 32  ->	   match (from_interval_type, to_interval_type):
 33  	       case (TimeIntervalType.PERIOD_BEGINNING, TimeIntervalType.PERIOD_ENDING):
 34  	           # shift time forward, 1am pd-beginning >> 2am pd-ending
 35  	           mult = 1
 36  	       case (TimeIntervalType.PERIOD_ENDING, TimeIntervalType.PERIOD_BEGINNING):
 37  	           # shift time backward
(Pdb) type(freq)
<class 'numpy.timedelta64'>
(Pdb) type(freq) is pd.Timedelta
False
```